src/pathfinder/water_regions.cpp

```diff
@@ -0,0 +1,405 @@
+			const TWaterRegionPatchLabel current_label = current_region.GetLabel(current_edge_tile);
+			if (current_label != water_region_patch.label) continue;
+
+			const TileIndex neighbor_edge_tile = GetEdgeTileCoordinate(nx, ny, opposite_side, x_or_y);
+			const TWaterRegionPatchLabel neighbor_label = neighboring_region.GetLabel(neighbor_edge_tile);
+			if (std::find(unique_labels.begin(), unique_labels.end(), neighbor_label) == unique_labels.end()) unique_labels.push_back(neighbor_label);
+		}
+		for (TWaterRegionPatchLabel unique_label : unique_labels) func(WaterRegionPatchDesc{ nx, ny, unique_label });
+	}
+}
+
+/**
+ * Calls the provided callback function on all accessible water region patches in
+ * each cardinal direction, plus any others that are reachable via aqueducts.
+ * @param water_region_patch Water patch within the water region to start searching from
+ * @param callback The function that will be called for each accessible water patch that is found
+ */
+void VisitWaterRegionPatchNeighbors(const WaterRegionPatchDesc& water_region_patch, TVisitWaterRegionPatchCallBack& callback)
+{
+	const WaterRegion &current_region = GetUpdatedWaterRegion(water_region_patch.x, water_region_patch.y);
+
+	/* Visit adjacent water region patches in each cardinal direction */
+	for (DiagDirection side = DIAGDIR_BEGIN; side < DIAGDIR_END; side++) VisitAdjacentWaterRegionPatchNeighbors(water_region_patch, side, callback);
+
+	/* Visit neigboring water patches accessible via cross-region aqueducts */
+	if (current_region.HasCrossRegionAqueducts()) {
+		for (const TileIndex tile : current_region) {
+			if (GetWaterRegionPatchInfo(tile) == water_region_patch && IsAqueductTile(tile)) {
+				const TileIndex other_end_tile = GetOtherBridgeEnd(tile);
+				if (GetWaterRegionIndex(tile) != GetWaterRegionIndex(other_end_tile)) callback(GetWaterRegionPatchInfo(other_end_tile));
+			}
+		}
+	}
+}
+
+void DEBUG_MarkRegion(int x, int y, int label, int color, bool override)  // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+{
+#ifndef NDEBUG
+	for (int dx = 0; dx < WATER_REGION_EDGE_LENGTH; ++dx) {
+		for (int dy = 0; dy < WATER_REGION_EDGE_LENGTH; ++dy) {
+			int tx = x * WATER_REGION_EDGE_LENGTH + dx;
+			int ty = y * WATER_REGION_EDGE_LENGTH + dy;
+			TileIndex tile = TileXY(tx, ty);
+
+			if (label < 0 || GetWaterRegionPatchInfo(tile).label == label)
+				DEBUG_addDrawRectangleInstruction(tile, color, override);  // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+		}
+	}
+#endif
+}
+
+void DEBUG_MarkRegion(TWaterRegionIndex region_index, int color, bool override)
+{
+	DEBUG_MarkRegion(GetWaterRegionX(region_index), GetWaterRegionY(region_index), -1, color, override);
+}
+
+std::vector<WaterRegionSaveLoadInfo> GetWaterRegionSaveLoadInfo()
+{
+	std::vector<WaterRegionSaveLoadInfo> result;
+	for (WaterRegion &region : _water_regions) result.push_back(WaterRegionSaveLoadInfo{ region.IsInitialized()});
```

Comment:
Inconsistent space in the `WaterRegionSaveLoadInfo`, but why not use `emplace_back`?
```suggestion
	for (WaterRegion &region : _water_regions) result.emplace_back(region.IsInitialized());
```